### PR TITLE
Minimal slashing protection

### PR DESCRIPTION
**Pull request by alonmuroch**

### CODE REVIEW

**Comment by roman-blox** [comment]
It should be against stage branch first

File: backend/path_sign_attestation_slashing_test.go [hunk 1]
```diff
@@ -2,7 +2,11 @@ package backend
 
 import (
 	"context"
+	"encoding/hex"
 	"fmt"
+	"github.com/bloxapp/eth2-key-manager/core"
```
Comment (roman-blox) [inline backend/path_sign_attestation_slashing_test.go:7]: imports should be with the following format
```
import (
    // built-in packages

    // third-party packages

    // internal packaged
)
```

File: backend/path_storage_slashing.go [hunk 2]
```diff
@@ -180,32 +180,35 @@ func loadAccountSlashingHistory(storage *store.HashicorpVaultStore, pubKey types
...
-		} else {
-			if attestation, err = storage.ListAllAttestations(pubKey); err != nil {
-				errs[0] = errors.Wrap(err, "failed to list all attestations data")
-				return
-			}
+		highestAtt = storage.RetrieveHighestAttestation(pubKey)
+		if highestAtt == nil {
+			errs[0] = errors.Errorf("highest attestation is nil")
 		}
+		//if err != nil {
```
Comment (roman-blox) [inline backend/path_storage_slashing.go:192]: Do we need to keep this?

File: backend/path_storage_slashing.go [hunk 3]
```diff
@@ -262,18 +265,22 @@ func storeAccountSlashingHistory(storage *store.HashicorpVaultStore, pubKey type
...
-				if err := storage.SaveAttestation(pubKey, attestation); err != nil {
-					attErrs[i] = errors.Wrapf(err, "failed to save attestation for slot %d", attestation.Slot)
-				}
-			}(i, attestation)
+		if err := storage.SaveHighestAttestation(pubKey, slashingHistory.HighestAttestation); err != nil {
+			attErrs[0] = errors.Wrapf(err, "failed to save attestation for slot %d", slashingHistory.HighestAttestation.Slot)
 		}
-		attWg.Wait()
+
+		//var attWg sync.WaitGroup
```
Comment (roman-blox) [inline backend/path_storage_slashing.go:272]: Do we need to keep this?

File: backend/store/slashing_store.go [hunk 4]
```diff
@@ -253,41 +149,41 @@ func (store *HashicorpVaultStore) ListAllProposals(key e2types.PublicKey) ([]*co
...
-	}
-
-	var ret core.BeaconAttestation
-	if err := json.Unmarshal(entry.Value, &ret); err != nil {
-		return nil, errors.Wrap(err, "failed to unmarshal beacon attestation object")
-	}
-
-	return &ret, nil
-}
+//// SaveLatestAttestation implements Storage interface.
```
Comment (roman-blox) [inline backend/store/slashing_store.go:152]: Do we need to keep this?

File: backend/store/store.go [hunk 5]
```diff
@@ -85,6 +89,15 @@ func FromInMemoryStore(ctx context.Context, inMem *in_memory.InMemStore, storage
 		}
 	}
 
+	// save highest att.
+	for _, acc := range wallet.Accounts() {
```
Comment (roman-blox) [inline backend/store/store.go:93]: I would parallel this in terms of performance improvements

File: e2e/tests/proposal_far_future_signing.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
+package tests
+
+import (
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/bloxapp/eth2-key-manager/core"
```
Comment (roman-blox) [inline e2e/tests/proposal_far_future_signing.go:8]: imports should be with the following format

```
import (
    // built-in packages

    // third-party packages

    // internal packaged
)
```

File: e2e/tests/attestation_no_slashing_data.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
+package tests
+
+import (
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/bloxapp/eth2-key-manager/core"
```
Comment (roman-blox) [inline e2e/tests/attestation_no_slashing_data.go:8]: imports should be with the following format

```
import (
    // built-in packages

    // third-party packages

    // internal packaged
)
```

File: e2e/tests/attestation_far_future_signing.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
+package tests
+
+import (
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/bloxapp/eth2-key-manager/core"
```
Comment (roman-blox) [inline e2e/tests/attestation_far_future_signing.go:8]: imports should be with the following format

```
import (
    // built-in packages

    // third-party packages

    // internal packaged
)
```

File: e2e/e2e_api_calls.go [hunk 9]
```diff
@@ -12,6 +12,8 @@ import (
 	"os"
 	"testing"
 
+	"github.com/bloxapp/eth2-key-manager/stores/in_memory"
```
Comment (roman-blox) [inline e2e/e2e_api_calls.go:15]: imports should be with the following format

```
import (
    // built-in packages

    // third-party packages

    // internal packaged
)
```

File: backend/store/store.go (hunk 5 above)
Comment (alonmuroch) [reply backend/store/store.go:93]: I think this is out of scope as we have other loops in the same func that are not async (been there before)

Comment (roman-blox) [reply backend/store/store.go:93]: up to you, but this is a potential bottleneck

**Pull request closed without merging**